src/java/main/org/apache/zookeeper/ClientCnxn.java

```diff
@@ -1107,13 +1098,23 @@ private void startConnect() throws IOException {
             String hostPort = addr.getHostString() + ":" + addr.getPort();
             MDC.put("myid", hostPort);
             setName(getName().replaceAll("\\(.*\\)", "(" + hostPort + ")"));
-            if (ZooKeeperSaslClient.isEnabled()) {
+            ZKClientConfig conf = zooKeeper.getClientConfig();
+            if (conf.isSaslClientEnabled()) {
                 try {
-                    String principalUserName = System.getProperty(
-                            ZK_SASL_CLIENT_USERNAME, "zookeeper");
-                    zooKeeperSaslClient =
-                        new ZooKeeperSaslClient(
-                                principalUserName+"/"+addr.getHostString());
+                    String principalUserName = conf.getProperty(ZKClientConfig.ZK_SASL_CLIENT_USERNAME,
+                            ZKClientConfig.ZK_SASL_CLIENT_USERNAME_DEFAULT);
+                    String serverPrincipal = principalUserName + "/" + addr.getHostString();
+                    // default is Client
+                    String loginContext = conf.getProperty(ZKClientConfig.LOGIN_CONTEXT_NAME_KEY,
+                            ZKClientConfig.LOGIN_CONTEXT_NAME_KEY_DEFAULT);
+                    if (null != zooKeeperSaslClient && zooKeeperSaslClient.getSaslState() != SaslState.FAILED) {
```

Comment:
ok, no problem lets crete ZooKeeperSaslClient every time. anyway major time is taken in createSaslClient() method, not in ZooKeeperSaslClient constuctor
So we can create the ZooKeeperSaslClient always